### DISCUSSION TITLE
Detecting changes in sub-properties

### DISCUSSION THREAD

**Discussion by tommillard** [lit/lit#2304] [Q&A] [answered]
I've made a simplified reduction of the issue I'm having. Some of the state of my component is held in an object. I'd like to be able to detect and react to a change in a sub-property with some fine control. Here's the example:

```
import { html, LitElement } from "lit";
import { customElement, state } from "lit/decorators.js";

@customElement("my-app")
export class App extends LitElement {
    @state({})
    data = {
        fruit: "apples",
        weather: "rain",
    };

    constructor() {
        super();
        setTimeout(() => {
            this.data.fruit = "bananas";
            this.data = { ...this.data };
        }, 3000);

        setTimeout(() => {
            this.data.weather = "snow";
            this.data = { ...this.data };
        }, 6000);
    }

    render() {
        return html`
            Weather: ${this.data.weather} Fruit: ${this.data.fruit}
        `;
    }

    shouldUpdate(changedProperties: any) {
        // I only want to update if the weather is changing, not the fruit
       console.log(
            `new weather is: ${changedProperties.get("data")?.weather}`
        );
        console.log(`current weather is: ${this.data.weather}`);
        console.log(`new fruit is: ${changedProperties.get("data")?.fruit}`);
        console.log(`current fruit is: ${this.data.fruit}`);
        console.log("");

        if (changedProperties.get("data")?.weather !== this.data.weather) {
            return true;
        } else {
            return false;
        }
    }
}
```

When shouldUpdate fires, the component's evaluation of the sub-property values has already updated. So I can't compare ```changedProperties.get("data")?.weather``` with ```this.data.weather``` to see if it has changed.

Does anyone know a new approach or fix? It works as expected if I split that state object into two separate state variables, but objects and properties works so much better in keeping code readable. Passing more than 4 or 5 properties into a component starts getting messy, in my opinion.

**Comment by kevinpschaaf** [comment, marked as answer]
This is similar in spirit to the discussion in #2319 so I'll post similar answers:

Lit has a simple dirty-checking model, where (by default) a strict equality check (`===`) is used to trigger the update cycle when a `@property`/`@state` field is set. In other words, the value of the field _must_ change to trigger an update, and the previous value is provided in the `changedProperties` passed to `willUpdate`/`updated`. 

This system is well-suited to "immutable data" patterns where root objects are created anew whenever a sub-property is changed (e.g. `this.value = {...this.value, someNewProp: 'foo'}`) rather than mutating the sub-property directly. So this is the first step: ensuring that immutable patterns are used rather than deep modifications of objects, to ensure Lit sees the change to the sub-property by means of the root property changing and triggers an update.

Second, because Lit doesn't store previous sub-property values or have a way of triggering an update based on a sub-property value changing, it doesn't have a way to tell you exactly what sub-property changed. So in order to know that `data.weather` changed (as opposed to e.g. `data.fruit`), you would need to allocate your own storage to hold the previous `data.weather` and do the dirty checking yourself. Something like:

```js
prevWeather = undefined;

shouldUpdate(changedProperties) {
  if (this.data?.weather !== this.prevWeather) {
    return true;
  } else {
    return false;
  }
}
```
